cloudinit/stages.py

```diff
@@ -248,17 +248,40 @@ def _get_data_source(self, existing):
             LOG.debug(myrep.description)
 
         if not ds:
-            util.del_file(self.paths.instance_link)
             (cfg_list, pkg_list) = self._get_datasources()
             # Deep copy so that user-data handlers can not modify
             # (which will affect user-data handlers down the line...)
-            (ds, dsname) = sources.find_source(self.cfg,
-                                               self.distro,
-                                               self.paths,
-                                               copy.deepcopy(self.ds_deps),
-                                               cfg_list,
-                                               pkg_list, self.reporter)
+            try:
+                (ds, dsname) = sources.find_source(self.cfg,
+                                                   self.distro,
+                                                   self.paths,
+                                                   copy.deepcopy(self.ds_deps),
+                                                   cfg_list,
+                                                   pkg_list, self.reporter)
+            except sources.DataSourceNotFoundException as e:
+                # Could not find a valid datasource.
+                if fallback:
+                    LOG.info("No datasource found - "
+                             "Trying to Load last one from cache")
+                    ds = self._restore_from_cache()
+                if not ds:
+                    raise e
+                dsname = ds.dsname
+                restored_iid = ds.metadata.get('instance-id', None)
+                for dsource in sources.list_sources(cfg_list,
+                                                    self.ds_deps,
+                                                    pkg_list):
+                    if (dsource.dsname == dsname and
+                       self.previous_iid() == restored_iid):
+                        break
+                else:
+                    LOG.info("Cached datasource '%s' doesn't match "
+                             "datasources enabled in the configuration",
+                             ds.dsname)
```

Comment:
Couple of thoughts here.  On modern instance which use ds-identify to detect the datasources, this ds list is just two elements, the detected DS and None;  However, on Xenial, for example, the list_sources() will be longer.  The else clause will only print out the last ds name it checked, so I'm wondering if it would be useful to users (or devs) who debug the scenario where the cached datasource doesn't match, to include the list of ds names that were checked.

Second though, in the case that the name matches, but the instance-id's don't match; we only see that the last ds didn't match the cached datasource;  I would think we'd want to indicate that we didn't use the cached datasource due to previous_iid() not matching restored_iid.

Reply:
> I'm wondering if it would be useful to users (or devs) who debug the scenario where the cached datasource doesn't match, to include the list of ds names that were checked.

Yes. I agree.

> Second though, in the case that the name matches, but the instance-id's don't match; we only see that the last ds didn't match the cached datasource; I would think we'd want to indicate that we didn't use the cached datasource due to previous_iid() not matching restored_iid.

In insight, I'm not sure this check is even useful.  Because we are not checking against the current instance ID (that would be ```check_instance_id()```'s job). So... is there really any case where the instance id in the pickled DataSource object would not match the instance id stored in ```previous-instance-id``` ? After all, if we move the volume to a new instance, we bring along both the pickled object and /var/lib/cloud/data.